frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.tsx

```diff
@@ -0,0 +1,151 @@
+}: {
+    elements: ElementType[]
+    highlight: boolean
+    editable: boolean
+    onChange: (i: number, s: string) => void
+}): JSX.Element {
+    return (
+        <>
+            {elements.map((element, index) => {
+                return (
+                    <SelectableElement
+                        key={`${element.tag_name}-${index}`}
+                        element={element}
+                        isDeepestChild={index === elements.length - 1}
+                        onChange={(s) => (editable ? onChange(index, s) : undefined)}
+                        readonly={!editable}
+                        indent={indent(index)}
+                        highlight={highlight}
+                    />
+                )
+            })}
+        </>
+    )
+}
+
+export function HtmlElementsDisplay({
+    elements: providedElements,
+    onChange,
+    highlight = true,
+    editable = false,
+    checkUniqueness = false,
+}: {
+    elements: ElementType[]
+    highlight?: boolean
+    editable?: boolean
+    checkUniqueness?: boolean
+    onChange?: (selector: string, isUnique?: boolean) => void
+}): JSX.Element {
+    let elements = [...(providedElements || [])].reverse()
+    elements = elements.slice(Math.max(elements.length - 10, 1))
+
+    const [selectors, setSelectors] = useState({} as Record<number, string>)
+    const [chosenSelector, setChosenSelector] = useState('')
+
+    const [selectorMatches, setSelectorMatches] = useState([] as HTMLElement[])
+
+    useEffect(() => {
+        let lastKey = -2
+        let builtSelector = ''
+
+        Object.keys(selectors)
+            .map((k) => Number.parseInt(k))
+            .sort()
+            .forEach((key) => {
+                const selector = selectors[key]
+                if (!!selector.trim().length) {
+                    if (lastKey === key - 1 && !!builtSelector.trim().length) {
+                        builtSelector += ` > ${selector}`
+                    } else {
+                        builtSelector += ` ${selector}`
+                    }
+                }
+                lastKey = key
+            })
+
+        builtSelector = !!builtSelector.trim().length ? builtSelector.trim() : 'no selectors chosen'
+        let selectorMatchCount = -1
+        if (builtSelector !== chosenSelector) {
+            if (checkUniqueness && !!builtSelector) {
+                try {
+                    const newSelectorMatches: HTMLElement[] = Array.from(document.querySelectorAll(builtSelector))
+                    selectorMatchCount = newSelectorMatches.length
+                    setSelectorMatches(newSelectorMatches)
+                } catch (e) {
+                    console.error(e)
+                    setSelectorMatches([])
+                }
+            }
+
+            setChosenSelector(builtSelector)
+            onChange?.(builtSelector, checkUniqueness ? selectorMatchCount === 1 : undefined)
+        }
+    }, [selectors, selectorMatches, checkUniqueness])
```

Comment:
This already looks like enough internal state to put in a logic (with a simple unique key)